contracts/FleekERC721.sol

```diff
@@ -267,6 +291,164 @@ contract FleekERC721 is ERC721, FleekAccessControl {
         emit NewTokenImage(tokenId, _tokenImage, msg.sender);
     }
 
+    /**
+     * @dev Add a new mirror registry for an app token.
+     * The mirror name should be a DNS or ENS url and it should be unique.
+     * Anyone can add a mirror but it should requires a payment.
+     *
+     * May emit a {NewMirror} event.
+     *
+     * Requirements:
+     *
+     * - the tokenId must be minted and valid.
+     *
+     * IMPORTANT: The payment is not set yet
+     */
+    function addMirror(uint256 tokenId, string memory mirrorName) public payable {
+        // require(msg.value == 0.1 ether, "You need to pay at least 0.1 ETH"); // TODO: define a minimum price
+        _requireMinted(tokenId);
+        require(_mirrors[mirrorName].owner == address(0), "Mirror already exists");
+
+        _mirrors[mirrorName] = Mirror(tokenId, _apps[tokenId].mirrors.length, 0, msg.sender);
+        _apps[tokenId].mirrors.push(mirrorName);
+
+        emit NewMirror(mirrorName, tokenId, msg.sender);
+    }
+
+    /**
+     * @dev Remove a mirror registry for an app token.
+     * Only the owner of the mirror can remove it.
+     * It will also remove the mirror from the app token mirrors list.
+     *
+     * May emit a {RemovedMirror} event.
+     *
+     * Requirements:
+     *
+     * - the mirror must exist.
+     */
+    function removeMirror(string memory mirrorName) public requireMirror(mirrorName) {
+        require(msg.sender == _mirrors[mirrorName].owner, "You are not the owner of this mirror");
+        App storage _app = _apps[_mirrors[mirrorName].tokenId];
+
+        // the index of the mirror to remove
+        uint256 indexToRemove = _mirrors[mirrorName].index;
+
+        // the last item is reposited in the index to remove
+        string memory lastMirror = _app.mirrors[_app.mirrors.length - 1];
+        _app.mirrors[indexToRemove] = lastMirror;
+        _mirrors[lastMirror].index = indexToRemove;
+
+        // remove the last item
+        _app.mirrors.pop();
+
+        delete _mirrors[mirrorName];
+        emit RemovedMirror(mirrorName, msg.sender);
+    }
+
+    /**
+     * @dev A view function to gether information about a mirror.
+     * It returns a JSON string representing the mirror information.
+     *
+     * Requirements:
+     *
+     * - the mirror must exist.
+     *
+     */
+    function mirror(string memory mirrorName) public view requireMirror(mirrorName) returns (string memory) {
```

Comment:
Maybe we call this getMirrorURI so its specific

Reply:
So it would be also delivered as base64? Here it is just returning a plain json string.

Reply:
Ah gotcha.  Yeah something else but still specific. Seeing `mirror` makes it seem like a constructor for mirror.  But just being picky we can change later.  Maybe something like getMirrorJson.